src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -141,17 +141,36 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
             var name = SymbolCompletionItem.GetSymbolName(item);
             var kind = SymbolCompletionItem.GetKind(item);
             var isGeneric = SymbolCompletionItem.GetSymbolIsGeneric(item);
-            var relatedDocumentIds = document.Project.Solution.GetRelatedDocumentIds(document.Id).Concat(document.Id);
             var options = document.Project.Solution.Workspace.Options;
-            var totalSymbols = await base.GetPerContextSymbolsAsync(document, position, options, relatedDocumentIds, preselect: false, cancellationToken: cancellationToken).ConfigureAwait(false);
-            foreach (var (documentId, syntaxContext, symbols) in totalSymbols)
+            var relatedDocumentIds = document.Project.Solution.GetRelatedDocumentIds(document.Id);
+            var typeConvertibilityCache = new Dictionary<ITypeSymbol, bool>(SymbolEqualityComparer.Default);
+
+            foreach (var relatedId in relatedDocumentIds)
             {
-                var bestSymbols = symbols.WhereAsArray(
-                    s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() > 0));
+                var relatedDocument = document.Project.Solution.GetRequiredDocument(relatedId);
+                var context = await CreateContextAsync(relatedDocument, position, cancellationToken).ConfigureAwait(false);
+                var symbols = await TryGetSymbolsForContextAsync(context, options, preselect: false, cancellationToken).ConfigureAwait(false);
 
-                if (bestSymbols.Any())
+                if (symbols.HasValue)
                 {
-                    return await SymbolCompletionItem.GetDescriptionAsync(item, bestSymbols, document, syntaxContext.SemanticModel, cancellationToken).ConfigureAwait(false);
+                    var bestSymbols = symbols.Value.Where(
+                        s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() > 0)).ToList();
+
+                    if (bestSymbols.Any())
+                    {
+                        if (IsTargetTypeCompletionFilterExperimentEnabled(document.Project.Solution.Workspace))
+                        {
+                            if (TryFindFirstSymbolMatchesTargetTypes(_ => context, bestSymbols, typeConvertibilityCache, out var index) && index > 0)
+                            {
+                                // This would ensure a symbol matches target types to be used for description (if any).
+                                var firstMatch = bestSymbols[index];
```

Comment:
It might be easier to understand if adding the comment saying the first symbol would be preferred and use as the description candidate